[GLIB] Allow developer permissions in bwrap sandbox with env var

TingPing opened the pull request:
#### 0908dc61bfc8dd1cf35640bfae4b7f33d087c636
<pre>
[GLIB] Allow developer permissions in bwrap sandbox with env var
<a href="https://bugs.webkit.org/show_bug.cgi?id=257809">https://bugs.webkit.org/show_bug.cgi?id=257809</a>

Reviewed by Michael Catanzaro.

This allows using tools like gdb, rr, or anything relying on ptrace
in the sandbox.

* Source/WebKit/UIProcess/Launcher/glib/BubblewrapLauncher.cpp:
(WebKit::setupSeccomp):
(WebKit::bubblewrapSpawn):

Canonical link: <a href="https://commits.webkit.org/265307@main">https://commits.webkit.org/265307@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/47b4fa4080fecd1d1c5d9f0a0c1f70900bed4b6c

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/9748 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/9998 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/10244 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/11404 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/9509 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/11984 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/9954 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/12432 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/9902 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/10730 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/8288 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/11563 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/8034 "Passed tests") | | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/16256 "Passed tests") | 
| | [❌ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/9133 "1 api test failed or timed out") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/9001 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/12334 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/9501 "Built successfully") | [❌ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/7746 "3 failures") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/8676 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/12902 "Built successfully") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/1178 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/9272 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

mcatanzaro reviewed (commented):
How would this work, though? How do you launch gdb inside the bwrap sandbox?

It's really unfortunate to have this gated behind a build guard. A runtime check would be much nicer and avoid the risk caused by divergence between user vs. developer builds. I can't believe I'm going to suggest this, but: using an environment variable would probably be appropriate?

mcatanzaro commented on Source/WebKit/UIProcess/Launcher/glib/BubblewrapLauncher.cpp:
```diff
@@ -442,6 +442,18 @@ static void bindV4l(Vector<CString>& args)
     }));
 }
 
+static bool enableDeveloperPermissions()
```
I would rename it from "developer" to "debug" everywhere

mcatanzaro commented on Source/WebKit/UIProcess/Launcher/glib/BubblewrapLauncher.cpp:
```diff
@@ -442,6 +442,18 @@ static void bindV4l(Vector<CString>& args)
+    static int enabled = -1;
+
+    if (enabled == -1) {
+        const char* env = g_getenv("WEBKIT_ENABLE_UNSAFE_DEVELOPER_PERMISSIONS_IN_SANDBOX");
```
Here too.

Do we really need to call it `UNSAFE`? I doubt anybody is going to try to using this on a permanent basis, like some people do to subvert the sandbox?

TingPing replied:
I only added it because you like scary names. Anything works for me.

mcatanzaro replied:
The scary name was needed for the environment variable that completely disables the sandbox because users and even applications were using it when they encountered bugs, without understanding the consequences. Understanding that ptrace is probably a sandbox escape(?) and this is probably really dangerous(?), I doubt that's going to be a problem here.

TingPing committed:
[GLIB] Allow developer permissions in bwrap sandbox with env var
https://bugs.webkit.org/show_bug.cgi?id=257809

Reviewed by Michael Catanzaro.

This allows using tools like gdb, rr, or anything relying on ptrace
in the sandbox.

* Source/WebKit/UIProcess/Launcher/glib/BubblewrapLauncher.cpp:
(WebKit::setupSeccomp):
(WebKit::bubblewrapSpawn):

Canonical link: https://commits.webkit.org/265307@main
